### PR TITLE
feat(gateways): enhance ingressgateways

### PR DESCRIPTION
**Pull request by robmcelvenny**
feat(gateways): enhance admin and passthrough ingress configurations

- Added a new security policy to the admin-ingressgateway for enhanced admin portal security.

- Implemented an optimized routing strategy for the passthrough gateway to improve load handling efficiency.

These changes improve the overall security posture and performance of the ingress gateways.

### CODE REVIEW

File: charts/addresspools/templates/ipaddresspool-admin-ingressgateway.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ apiVersion: metallb.io/v1beta1
 kind: IPAddressPool
 metadata:
   name: admin-ingressgateway
-  namespace: metallb-system
```
Comment (andrewg-xyz) [inline charts/addresspools/templates/ipaddresspool-admin-ingressgateway.yaml:6]: namespace should probably be a variable, with default "metallb-system" to ensure backwards compatibility

(true for all changes here)

Comment (blancharda) [reply charts/addresspools/templates/ipaddresspool-admin-ingressgateway.yaml:6]: I don't think the namespace for the `IPAddressPool` itself needs to change (they can still live in `metallb-system`). In fact, it should probably be set to `{{ .Release.Namespace }} ` and handled by helm.

But I agree with adding templated vars for the selector namespace and names for added flexibility.
Maybe something like this in the default values with overrides available via zarf:
```
default:
  ipAddressPool: ""

adminIngress:
  ipAddressPool: ""
  namespace: "istio-admin-gateway
  label:
    - app=admin-ingressgateway
    
adminIngress:
  ipAddressPool: ""
  namespace: "istio-tenant-gateway
  label:
    - app=tenant-ingressgateway
    
passthroughIngress:
  ipAddressPool: ""
  namespace: "istio-passthrough-gateway
  label:
    - app=passthrough-ingressgateway
 ```

File: charts/addresspools/templates/ipaddresspool-admin-ingressgateway.yaml [hunk 2]
```diff
@@ -3,7 +3,7 @@ apiVersion: metallb.io/v1beta1
...
+  namespace: istio-admin-gateway
 spec:
   addresses:
     - {{ .Values.ipAddressAdminIngressgateway }}/32
```
Comment (blancharda) [inline charts/addresspools/templates/ipaddresspool-admin-ingressgateway.yaml:9]: I'm not sure what the expected address format is here.. but this would likely need to be updated if we adopt my suggested values above.

File: charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml
Comment (blancharda) [inline charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml:1]: I think `spec.serviceAllocation.namespaces` would need to be updated here as well

**Review by blancharda — CHANGES_REQUESTED** [review]
See comments.

Suggest:
* replace namespace with chart's `{{ Release.Namespace }}` to align with typical helm practice
* reorganize the chart values into sections per gateway
* template more variables for flexibility and configuration

File: charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml
Comment (robmcelvenny) [reply charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml:1]: @blancharda I figured it out, this works and provides the IP addresses that can be defined in a singular way (rather than a range )

ipaddresspool-keycloak-ingressgateway.yaml
ipaddresspool-tenant-ingressgateway.yaml
ipaddresspool-admin-ingressgateway.yaml 

don't appear to be necessary, or maybe there is a better way to get it to work, it appears IPAddressPool only accepts the following as valid IP addresses, each one was able to be picked up and assigned with istio.
```
{{- if .Values.ipAddressAdminIngressgateway }}
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: default
  namespace: metallb-system
spec:
  addresses:
    - {{ .Values.ipAddressAdminIngressgateway }}/32 
    - {{ .Values.ipAddressTenantIngressgateway }}/32
    - {{ .Values.ipAddressKeycloakIngressgateway }}/32
{{- end }}
```

Comment (blancharda) [reply charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml:1]: I'm not positive, but I think this will result in the 3 load balancers (potentially) switching between those 3 addresses, which is no different than the current behavior with the range in the default pool.

The purpose of the individual `IPAddressPool` objects is to provide a specific IP (or range) to a specific ingressgateway, identified by a label selector and namespace.

I won't have a chance to get back to this today, but can take a closer look next week 🤔

Comment (robmcelvenny) [reply charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml:1]: Ok, I tried again, I removed spec.serviceAllocation from each template and the IP's got allocated correctly, so the serviceAllocation isn't working or configured properly.

```
  serviceAllocation:
    priority: 50
    namespaces:
      - istio-system
    serviceSelectors:
      - matchLabels:
          app: admin-ingressgateway
```

File: charts/addresspools/templates/ipaddresspool-admin-ingressgateway.yaml (hunk 1 above)
Comment (robmcelvenny) [reply charts/addresspools/templates/ipaddresspool-admin-ingressgateway.yaml:6]: Alright, I updated and tested it out and appears to be working good now.

File: charts/addresspools/templates/ipaddresspool-default.yaml
Comment (blancharda) [inline charts/addresspools/templates/ipaddresspool-default.yaml:1]: I think the if statement at the top needs to be updated

File: charts/addresspools/templates/ipaddresspool-default.yaml [hunk 3]
```diff
@@ -1,10 +1,10 @@
-{{- if .Values.ipAddressPool }}
+{{- if .Values.ipAddressPoolCIDR }}
```
Comment (mjnagel) [inline charts/addresspools/templates/ipaddresspool-default.yaml:1]: This doesn't match the default values you have set.
```suggestion
{{- if .Values.default.ipAddressPoolCIDR }}
```

File: charts/addresspools/templates/ipaddresspool-default.yaml [hunk 4]
```diff
@@ -1,10 +1,10 @@
...
 apiVersion: metallb.io/v1beta1
 kind: IPAddressPool
 metadata:
   name: default
   namespace: metallb-system
 spec:
   addresses:
-    - {{ .Values.ipAddressPool }}
+    - {{ .Values.ipAddressPoolCIDR }}
```
Comment (mjnagel) [inline charts/addresspools/templates/ipaddresspool-default.yaml:9]:
```suggestion
    - {{ .Values.default.ipAddressPoolCIDR }}
```

File: zarf.yaml [hunk 5]
```diff
@@ -16,7 +16,7 @@ variables:
     description: When using this package in concert with DUBBD, the IP address of the admin-ingressgateway service
```
Comment (mjnagel) [inline zarf.yaml:16]: think we can update these comments to reference UDS Core instead of DUBBD/IDAM.

File: charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml [hunk 6]
```diff
@@ -1,17 +1,25 @@
...
-    - {{ .Values.ipAddressTenantIngressgateway }}/32
+    {{- if not (empty .Values.tenantIngress.ipAddressPool) }}
+    {{- if (kindIs "string" .Values.tenantIngress.ipAddressPool) }}
+    - {{ .Values.tenantIngress.ipAddressPool }}/32
+    {{- else }}
+    {{- range .Values.tenantIngress.ipAddressPool }}
+    - {{ . }}/32
+    {{- end }}
+    {{- end }}
+    {{- end }}
```
Comment (mjnagel) [inline charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml:17]: I might just be missing the intent here, but I don't think you need this template to be complicated. If someone wants a specific IP they would just pass in a single IP value - don't think there's a strong use case for passing in multiple values for a single gateway?

File: charts/addresspools/Chart.yaml [hunk 7]
```diff
@@ -17,5 +17,5 @@ type: application
 # Versions are expected to follow Semantic Versioning (https://semver.org/)
 # This version should stay in sync with the version in the zarf.yaml file.
 # x-release-please-start-version
-version: "0.0.5"
+version: "0.0.6"
```
Comment (mjnagel) [inline charts/addresspools/Chart.yaml:20]: nit: anywhere you have bumped the version it can be reverted since release-please will manage that for you in the release PR.

**Review by mjnagel — CHANGES_REQUESTED** [review]

File: charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml (hunk 6 above)
Comment (blancharda) [reply charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml:17]: I may have overcomplicated the suggestion. The goal was to keep it as flexible as possible, but I agree, providing multiple IPs per gateway is probably a pretty unlikely edge case.

Comment (robmcelvenny) [reply charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml]: @blancharda np! I already made / pushed the changes. 👍 was a fun little test, and I'm sure there might be a few other ways to do it. Learned a lot though.

File: charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml [hunk 8]
```diff
@@ -1,17 +1,25 @@
...
   serviceAllocation:
     priority: 50
     namespaces:
-      - istio-system
+      - {{ .Values.tenantIngress.namespace }}
     serviceSelectors:
       - matchLabels:
-          app: tenant-ingressgateway
+          {{ .Values.tenantIngress.matchLabel }}
```
[the remark this answers is not in the document]
Comment (robmcelvenny) [reply charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml:16]: Might need to test this more, but I did change to app and tested it, how can I verify if it aplied / matched label correctly?

File: README.md [hunk 9]
```diff
@@ -34,20 +34,20 @@ $ zarf package deploy oci://ghcr.io/defenseunicorns/packages/metallb:<version> \
     --confirm
 ```
 
-Or, in the case of using this package in concert with DUBBD and/or IDAM package:
+Or, in the case of using this package in concert with UDS Core and/or IDAM package:
```
Comment (mjnagel) [inline README.md:37]:
```suggestion
Or, in the case of using this package in concert with UDS Core package:
```

IDAM package would only be used with DUBBD so I think we could update the doc to just ref Core.

File: README.md [hunk 10]
```diff
@@ -34,20 +34,20 @@ $ zarf package deploy oci://ghcr.io/defenseunicorns/packages/metallb:<version> \
...
-    --set IP_ADDRESS_KEYCLOAK_INGRESSGATEWAY=10.0.0.34 \
+    --set IP_ADDRESS_PASSTHROUGH_INGRESSGATEWAY=10.0.0.34 \
     --confirm
 ```
 
 > Notes:
 >   - The IP addresses used here are placeholders. You can use whatever values you want that work for your environment.
 >   - Package versions can be found [here](https://github.com/defenseunicorns/uds-package-metallb/pkgs/container/packages%2Fmetallb)
->   - If you won't be using Keycloak, you can omit setting the `IP_ADDRESS_KEYCLOAK_INGRESSGATEWAY` variable. The package is smart enough to not create the associated IPAddressPool if the variable is not set.
+>   - If you won't be using Keycloak, you can omit setting the `IP_ADDRESS_PASSTHROUGH_INGRESSGATEWAY` variable. The package is smart enough to not create the associated IPAddressPool if the variable is not set.
```
Comment (mjnagel) [inline README.md:50]: Maybe also drop this - keycloak is required as part of core so it's not really optional unless using DUBBD.
```suggestion
```

File: charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (mjnagel) [reply charts/addresspools/templates/ipaddresspool-tenant-ingressgateway.yaml:16]: I helm templated this out and it was getting set to nothing. I would recommend doing this to validate:
```console
helm template charts/addresspools -f values/addresspools-values.yaml | less
```

If you update this line to:
```yaml
          {{- toYaml .Values.tenantIngress.matchLabels | nindent 10 }}
```

It will template out as expected. You would need the same for the other two addresspools.

**Review by mjnagel — CHANGES_REQUESTED** [review]
From testing out the helm template as well it looked like the conditionals on all the `l2advertisement-*` files will need to be updated - they weren't templating out as expected.

**Review by mjnagel — COMMENTED** [review]
Last issue I see - https://github.com/defenseunicorns/uds-capability-metallb/blob/update-gateway-namespaces/charts/addresspools/templates/l2advertisement-default.yaml#L1 - this conditional needs to be updated to `{{- if .Values.default.ipAddressPoolCIDR }}` I believe. Otherwise all seems to work as expected!

**Comment by mjnagel** [comment]
@RothAndrew / @blancharda I believe you two are using this in environments currently - just wanted to add your review request to make sure everything here is okay. This does switch the default to assume UDS Core, but otherwise maintain the same configurability. To use with DUBBD you would need to leverage bundle overrides to the helm values for `matchLabels`.

Pending the last conditional update this LGTM from my perspective.

**Review by mjnagel — APPROVED** [review]

**Review by RothAndrew — APPROVED** [review]

**Review by blancharda — APPROVED** [review]
lgtm!
